### DISCUSSION TITLE
Stop characters from being wrapped to the beginning of the line

### DISCUSSION THREAD

**Discussion by Casper-Guo** [microsoft/terminal#16724] [Q&A] [answered]
My terminal wraps a line that exceeds the character limit to the beginning of the same line and obfuscate the prompt:

![image](https://github.com/microsoft/terminal/assets/89810860/e171cccd-7370-453f-9a91-e7dfd0c3e0df)

What are the settings needed to prevent this and wrap to the next line (or expand per line character limit) instead?

**Comment by DHowett** [comment]
This looks like an issue with your shell. Can you share your custom `PS1`?

**Reply by Casper-Guo** [reply]
This is the part of my `.bashrc` that implements the custom `PS1`

```
function colorized_prompt {
    if [ $? -ne 0 ]; then
        printf "\e[31m$\e[m"
    else
        printf "\e[32m$\e[m"
    fi
}

function git_branch {
    git status 2>/dev/null | grep 'On branch' | sed 's/On branch / => /'
}

export PS1="\e[36m[\d \t]\e[m \e[32m\u\e[m\e[33m@\e[m\e[31m\h:\e[m\e[36m\w\e[m \$(colorized_prompt)\e[33m\$(git_branch)\e[m "
```

**Reply by DHowett** [reply, marked as answer]
Thanks!

Alright, so, one of the important things in a bash prompt is to enclose the non-printable escape sequences in `\[` and `\]`. That way, bash doesn't consider them when it measures how long the prompt is.

Try this out! It may help, or it may (depending on how well I remember bash...) break terribly.

```bash
function colorize_prompt {
    if [ $? -ne 0 ]; then
        printf "\e[31m"
    else
        printf "\e[32m"
    fi
}

function git_branch {
    git status 2>/dev/null | grep 'On branch' | sed 's/On branch / => /'
}

export PS1="\[\e[36m\][\d \t]\[\e[m\] \[\e[32m\]\u\[\e[m\e[33m\]@\[\e[m\e[31m\]\h:\[\e[m\e[36m\]\w\[\e[m\] \[\$(colorize_prompt)\]$\[\em\e[33m\]\$(git_branch)\[\e[m\] "
```

Here's what I did:

- Every place there was a `\e...m` I put `\[` before and `\]` after it.
- I couldn't do that for `colorized_prompt` because emitting the prompt escape wrapping tokens `\[` and `\]` inside a called function gets weird (and I forgot how to do it), so I made it _only emit the color SGR_ and then wrapped the function call in `\[` and `\]`. The `$` from inside colorize went outside the function call (e.g. `\[\$(colorize_prompt)\]` plus `$` plus `\[\e[m\]`)

There's a little more line noise, but I can now edit wrapped commands _and wrap commands_ properly!

**Reply by Casper-Guo** [reply]
Thanks, work as expected!

A tangential question is why are the brackets mismatched when specifying SGR such as `\e[33m\]@\[\e[m`? It seems like the last bracket is never closed

**Reply by DHowett** [reply]
Ah! When there were two escapes next to eachother, I just wrapped them up in the same `\[` `\]` pair:

<img width="973" alt="image" src="https://github.com/microsoft/terminal/assets/189190/e41ba1fd-496c-433b-b44d-d9802ecd5889">
